Rithika taking over for Aditya-feat: Implement Paid Labor Cost API endpoint with comprehensive filtering and validation

Aditya-gam opened the pull request:
# Description

Implements a complete backend API for the Paid Labor Cost grouped-bar graph feature. The controller previously returned an empty array as a stub. This PR provides a fully functional `GET /api/labor-cost` endpoint supporting date filtering, project filtering, task filtering, and proper data formatting with comprehensive validation and error handling.
<img width="677" height="273" alt="IssueDescription" src="https://github.com/user-attachments/assets/cb87c7cc-078a-45d2-a5f5-5892168d50b9" />


---
## Related PRs (if any):

- Related to old PRs: [#3420](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/3420) + [#1451](https://github.com/OneCommunityGlobal/HGNRest/pull/1451)
- Frontend Pair for this PR: [PR #4496](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/4496)
---
## Main changes explained:

### Created/Updated Files:

1. `src/controllers/bmdashboard/bmPaidLaborCostController.js`
	Complete controller implementation with:
	- **Helper functions:** `looksLikeJson()`, `parseArrayParam()`, `parseDateRangeParam()`, `isValidDateValue()`
	- **Parameter extraction & validation:** Extracts and validates `projects`, `tasks`, `date_range` from query params
	- **MongoDB query building:** Supports partial date ranges, project/task filters using the `$in` operator
	- **Database querying:** Uses `LaborCost.find()` with `.lean()` and `.sort({ date: 1 })`
	- **Response formatting:** Maps DB fields (`project_name` → `project`) and calculates `totalCost`
	- **Error handling:** Distinguishes database vs unexpected errors, logs with context
	- **Edge cases:** Handles null params, empty results, invalid dates (e.g., "2025-13-01"), UTC timezone handling
	
2. `src/routes/bmdashboard/bmPaidLaborCostRouter.js`
	- Changed route method from `POST` to `GET` to align with frontend requirements.
	
3. `src/controllers/bmdashboard/__tests__/bmPaidLaborCostController.test.js`
	119 unit tests covering:
	- Helper functions (59 tests): JSON detection, array parsing, date validation
	- Controller logic (60 tests): Parameter validation, query building, response formatting, error handling, edge cases
	- **Coverage:** 97% (above 90% target)
	
4. `src/__tests__/integration/bmPaidLaborCostRoutes.integration.test.js` (+113 lines)
Integration tests for route configuration, authentication (401 without auth), and GET method verification.
	
5. `package.json`
	- Added `@aws-sdk/client-s3` dependency for integration tests.
---
### Key Implementation Details:

**API Endpoint:** `GET /api/labor-cost.`

**Request Parameters (query string):**
- `projects`: Optional array (empty = all projects)
- `tasks`: Optional array (empty = all tasks)
- `date_range`: Optional object with `start_date` and/or `end_date` (null = all dates)

**Response Structure:**
```json
{
	"totalCost": 15000,
	"data": [
		{ "project": "Project A", "task": "Task 1", "date": "2025-04-01T00:00:00.000Z", "cost": 5000 }
	]
}
```

**Key Features:**
- **Partial date ranges:** Supports only start_date, only end_date, or both
- **UTC timezone handling:** Uses UTC methods for consistent date filtering
- **Flexible parameter parsing:** Accepts JSON strings, comma-separated values, or parsed arrays
- **Comprehensive validation:** Date format (ISO 8601), date values (catches "2025-13-01"), date range logic, array formats
- **Error responses:** 400 (invalid params), 422 (invalid dates), 500 (server errors)
## How to test:

1. Check into branch: `Aditya-fix/implement-paid-labor-cost-controller.`
2. Reinstall dependencies and clean cache using `rm -rf node_modules package-lock.json && npm cache clean --force.`
3. Run `npm install` to install dependencies, then start the backend locally (`npm start` or `npm run dev`)
4. **Test API Endpoint `GET /api/labor-cost`:**

**Basic functionality:**
```bash
# Get all data
curl "http://localhost:3000/api/labor-cost"

# Filter by date range
curl "http://localhost:3000/api/labor-cost?date_range={\"start_date\":\"2025-04-01\",\"end_date\":\"2025-04-30\"}"

# Filter by projects and tasks
curl "http://localhost:3000/api/labor-cost?projects=[\"Project A\"]&tasks=[\"Task 1\"]"
```

**Edge cases:**
```bash
# Partial date range (only start_date)
curl "http://localhost:3000/api/labor-cost?date_range={\"start_date\":\"2025-04-01\",\"end_date\":null}"

# Empty results (future dates)
curl "http://localhost:3000/api/labor-cost?date_range={\"start_date\":\"2030-01-01\",\"end_date\":\"2030-12-31\"}"

# Invalid date format (should return 422)
curl "http://localhost:3000/api/labor-cost?date_range={\"start_date\":\"2025-13-01\",\"end_date\":\"2025-04-30\"}"
```

5. **Run test suite:**
```bash
# Unit tests (119 tests, 97% coverage)
npm test -- src/controllers/bmdashboard/__tests__/bmPaidLaborCostController.test.js

# Integration tests
npm test -- src/__tests__/integration/bmPaidLaborCostRoutes.integration.test.js
```
## Screenshots or videos of changes:
  

https://github.com/user-attachments/assets/b7a5e90f-041a-4e7e-8f60-00f4a09321dd

---
## Note:

- **Database Changes:** Added some mock data to the dev database, so testing this endpoint is easier.
- **Performance:** Uses `.lean()` for better performance. Consider adding indexes on `date`, `project_name`, and `task` if the dataset is very large
- **Validation:** Comprehensive validation for dates (ISO 8601, invalid months/days), arrays (format detection), and date ranges
- **Error Handling:** User-friendly messages with error codes. Internal details not exposed. Errors logged with full context
- **Testing:** 97% coverage (119 unit tests + integration tests). All edge cases and error conditions covered

Aditya-gam committed:
feat: add model imports and error handling structure to paid labor cost controller

- Import LaborCost model for database queries
- Import logger utility for error tracking
- Wrap controller function in try-catch block for error handling
- Set up foundation for complete controller implementation

Aditya-gam committed:
feat: add parameter extraction and validation for paid labor cost controller

- Extract projects, tasks, and date_range from query parameters
- Add helper functions to parse arrays and date_range objects from query strings
- Validate projects and tasks are arrays with string values
- Validate date_range is object or null
- Validate ISO 8601 date format for start_date and end_date
- Validate date range logic (start_date <= end_date)
- Return appropriate error responses (400, 422) for validation failures

Aditya-gam committed:
feat: build MongoDB query filters for paid labor cost controller

- Build date filter with support for null date_range, partial ranges (only start or end), and full ranges
- Use MongoDB operators $gte and $lte for date filtering with inclusive boundaries
- Build project filter using $in operator when projects array has values
- Build task filter using $in operator when tasks array has values
- Combine all filters into single query object with AND logic
- Convert date strings to Date objects and set appropriate time boundaries for inclusive filtering

Aditya-gam committed:
feat: query database and retrieve labor cost records

- Execute MongoDB query using LaborCost.find() with combined filters
- Sort results chronologically by date using .sort({ date: 1 })
- Use .lean() for better query performance
- Return individual records without aggregation (frontend handles grouping)
- Database errors handled gracefully by existing try-catch block

Aditya-gam committed:
feat: format response data for paid labor cost controller

- Map database fields to frontend structure (project_name → project)
- Convert date objects to ISO 8601 string format
- Ensure cost values are numbers (not strings)
- Calculate totalCost as sum of all cost values
- Return response with totalCost and data array structure
- Handle empty results case (totalCost = 0)

Aditya-gam committed:
feat: construct final response object for paid labor cost controller

- Explicitly construct response object with totalCost and data fields
- Return 200 OK status code on successful response
- Handle empty results case (totalCost: 0, data: [])
- Ensure response is valid JSON using res.json() method
- Add clear comments documenting response structure

Aditya-gam committed:
feat: enhance error handling for paid labor cost controller

- Distinguish between database errors and unexpected errors
- Add error context logging with request parameters (query, method, URL)
- Return user-friendly error messages without exposing internal details
- Handle MongoDB/database connection errors with specific error code
- Maintain existing validation error handling (400, 422 status codes)
- Log errors with sufficient detail for debugging using logger utility

Aditya-gam committed:
feat: enhance edge case handling for paid labor cost controller

- Explicitly handle null/undefined request query parameters
- Enhance date validation to catch invalid months/days (e.g., "2025-13-01")
- Document empty results handling (valid case, not error)
- Document partial date range support (open-ended ranges)
- Add comments clarifying edge case behaviors
- Ensure dates far in past/future return empty results gracefully
- Validate date components match parsed values to prevent silent adjustments

Aditya-gam committed:
fix: change route method from POST to GET for paid labor cost endpoint

- Update route to use GET method as per frontend requirements
- Controller already reads from req.query which is appropriate for GET requests

Aditya-gam committed:
fix: improve date validation and parameter parsing for paid labor cost controller

- Fix date validation to use UTC methods for accurate date-only string comparison
- Add validation to detect JSON parsing failures for date_range parameter
- Add validation to ensure date values are strings (not numbers or other types)
- Improve error messages with examples for invalid date_range JSON format
- Add comments clarifying UTC method usage for date-only string parsing

Aditya-gam committed:
feat: enhance array parameter validation for paid labor cost controller

- Add validation to detect JSON objects provided instead of arrays
- Add validation to detect mixed formats (JSON objects/arrays with plain strings)
- Add helper function to detect JSON-like strings in comma-separated values
- Return descriptive error messages for invalid array parameter formats
- Improve validation for both projects and tasks array parameters

Aditya-gam committed:
fix: use UTC methods for date boundary setting in paid labor cost controller

- Change setHours to setUTCHours for start date boundary
- Change setHours to setUTCHours for end date boundary
- Add comments clarifying UTC method usage for consistent timezone handling
- Ensure date filters work correctly across different timezones

Aditya-gam committed:
test: add unit tests for paid labor cost controller helper functions

- Export helper functions via testExports for direct unit testing
- Add comprehensive tests for looksLikeJson() (12 test cases)
- Add comprehensive tests for parseArrayParam() (15 test cases)
- Add comprehensive tests for parseDateRangeParam() (10 test cases)
- Add comprehensive tests for isValidDateValue() (22 test cases)
- All 59 test cases passing, covering edge cases and error conditions

Aditya-gam committed:
test: add unit tests for paid labor cost controller logic

- Add comprehensive tests for parameter validation (projects, tasks, date_range)
- Add tests for MongoDB query building with various filter combinations
- Add tests for response formatting and data transformation
- Add tests for error handling (database errors and unexpected errors)
- Achieve 97% code coverage with 105 test cases
- All tests passing, covering edge cases and error conditions

Aditya-gam committed:
test: add integration tests for paid labor cost routes

- Add authorization tests for GET /api/labor-cost (401 without auth)
- Add route functionality tests (POST method, route existence)
- Follow same pattern as existing integration tests
- Tests verify route configuration and authentication requirements
- Install @aws-sdk/client-s3 dependency required for integration tests

Aditya-gam committed:
test: add edge case tests for paid labor cost controller

- Add boundary condition tests (empty strings, long names, Unicode, special chars)
- Add tests for large arrays (100+ projects)
- Add tests for dates far in past/future
- Add request object edge case tests (missing/null/empty query)
- All 115 tests passing with 97% coverage maintained

Aditya-gam committed:
test: add documentation and optimize paid labor cost tests

- Add comprehensive JSDoc comments and section headers to test files
- Consolidate request object edge case tests into parameterized test
- Maintain 97% coverage (well above 90% target)
- All 119 tests passing
- Improve test file readability and maintainability

Aditya-gam committed:
refactor: extract validation and parsing logic into helper functions for paid labor cost controller

Aditya-gam committed:
fix: Move @aws-sdk/client-s3 from devDependencies to dependencies

Anusha-Gali reviewed (approved):
Hi Aditya,

I have reviewed your PR locally and the backend API works as per requirement.

<img width="1267" height="763" alt="Image" src="https://github.com/user-attachments/assets/e2b0379e-34b9-485a-a1a4-dab7acdf3615" />
<img width="1267" height="763" alt="Image" src="https://github.com/user-attachments/assets/b3523fd6-8d4e-45ca-83fa-d5917f603f39" />
<img width="1267" height="763" alt="Image" src="https://github.com/user-attachments/assets/b7fc5eca-fac2-4568-a5dc-293c9d4157b2" />
<img width="1267" height="763" alt="Image" src="https://github.com/user-attachments/assets/f2b7a4a9-0f3b-4de0-bf42-896d1e5a14ed" />
<img width="1267" height="763" alt="Image" src="https://github.com/user-attachments/assets/2b4acc30-af9a-4d91-8eff-94e9b7f64c70" />
<img width="1267" height="763" alt="Image" src="https://github.com/user-attachments/assets/59f5fb85-e713-41dc-be46-c8f6462c0c02" />
<img width="933" height="763" alt="Image" src="https://github.com/user-attachments/assets/b7e99aea-9ccc-4661-9935-518189239798" />

Aditya-gam committed:
Merge branch 'development' into Aditya-fix/implement-paid-labor-cost-controller

Aditya-gam committed:
Merge branch 'development' into Aditya-fix/implement-paid-labor-cost-controller

Aditya-gam committed:
chore: align package-lock.json with development

Aditya-gam committed:
fix(security): build labor-cost query from validated params only

Extract findLaborCostByValidatedFilter() so the MongoDB query is not
constructed directly from req.query. Satisfies SonarQube rule on
user-controlled data in database queries.

Made-with: Cursor

Aditya-gam committed:
test: reduce duplication in paid labor cost controller tests

Refactor repetitive getLaborCost test setup/assertions into shared helpers and
parameterized cases to satisfy SonarQube duplicated-lines requirements.

Made-with: Cursor

Aditya-gam committed:
fix(security): sanitize project/task arrays to prevent NoSQL injection

Add sanitizeStringArrayForQuery() to filter MongoDB operator strings ($ prefix)
from user-provided project and task values before building query filters.

Made-with: Cursor

Aditya-gam committed:
test: add sanitizeStringArrayForQuery tests and reduce helper duplication

Convert looksLikeJson and isValidDateValue to it.each, add sanitizer unit
tests, and consolidate shared helpers.

Made-with: Cursor

rithika-paii committed:
fix: resolve merge conflicts with development branch
